Add hsm_mode: Half-Sample Mode for continuous data

anurag-mds opened the pull request:
<h2>Summary</h2>
This PR adds hsm_mode(), an implementation of the half-sample mode (HSM) which is a robust estimator of the mode for continuous distributions.

It is introduced as a separate function ( not an overload of mode() ) to preserve existing behaviour while providing a statistically meaningful alternative for continuous data 

This addresses and closes issue #957.
<hr>
<h2>Motivation</h2>
StatsBase.mode() is frequency-based and works well for discrete data. For continuous distributions, however, samples are usually unique, which makes frequency counts unstable and highly variable in practice

Issue #957 documents this behaviour, particularly for heavy-tailed distributions, where mode() can show extreme variance. 

This PR provides an estimator designed specifically for continuous data
<hr>
<h2>Approach</h2>
hsm_mode() implements the standard half-sample method described in the literature:

- Non-finite values (NaN, Inf) are filtered

- The data are sorted

- The algorithm repeatedly selects the contiguous half-sample with the smallest width

- Once ≤ 2 points remain, the midpoint of the final interval is returned

- The midpoint may not be a sample value, but provides a stable estimate of the location of highest density.

- Time complexity is dominated by sorting (O(n log n)); space complexity is O(n).

- After sorting, the contraction loop operates on SubArray views to avoid allocations.
<hr>
<h2>API Design</h2>

The estimator is exposed as a new function:

**hsm_mode(x::AbstractVector{T}) where T<:Real**


It is NOT added as an overload of mode() in order to:

- avoid changing existing semantics

- clearly distinguish frequency-based and density-based estimation

- let users choose the appropriate method explicitly

The **return type** is an **AbstractFloat**, promoted from the input element type (e.g. integers → Float64, Float32 → Float32).
<hr>
<h2>Testing and Documentation</h2>

Tests cover basic correctness, edge cases, robustness to outliers, handling of non-finite values, and type behavior. All tests pass.

The docstring explains intended use cases, compares with mode(), documents complexity, provides examples, and cites the relevant literature.
<hr>
<h2>References</h2>

<a href="https://www.tandfonline.com/doi/abs/10.1080/01621459.1974.10480246">Robertson & Cryer (1974), JASA</a>

<a href="https://www.sciencedirect.com/science/article/abs/pii/S0167947305001581">Bickel & Fruehwirth (2006), CSDA</a>
<hr>
<h2>Notes</h2>

This PR is intentionally small and focused. Extensions such as weighted HSM or support for missing values are left for future work.

<h3><u><i>I have attached images showing that the tests are passing, and I would like to know whether I should address the existing warnings in the codebase.</i></u></h3>

<b>Images: </b><br>
<img width="1900" height="1100" alt="image" src="https://github.com/user-attachments/assets/9b02bb37-4e72-4ea0-9166-e602f5024dd2" /><br>
<img width="1919" height="1128" alt="image" src="https://github.com/user-attachments/assets/4adb0560-07c3-41c7-8572-2615eaff1b85" /><br>
<img width="1917" height="1130" alt="image" src="https://github.com/user-attachments/assets/c6622e15-82f1-425a-903b-693a987a2ac5" /><br>
<img width="1918" height="1126" alt="image" src="https://github.com/user-attachments/assets/479c3f5c-e1fa-44f4-affd-783ec5072e9f" /><br>
<img width="1917" height="1114" alt="image" src="https://github.com/user-attachments/assets/409575ca-df86-492b-a3d5-3c1c93dbdbfb" /><br>
<img width="1445" height="695" alt="image" src="https://github.com/user-attachments/assets/008af2e7-910a-4f58-9106-3caa5608b108" />


_**Feedback on naming or API placement is welcome.**_

Commit:
Add half-sample mode via mode(x; method=:halfsample)

devmotion commented:
Was this PR AI-generated?

devmotion commented:
Was this PR AI-generated?

devmotion commented:
Was this PR AI-generated?

anurag-mds commented:
The implementation, test and commits are mine. I reviewed existing StatsBase prs to match project conventions, and I used AI assistance only for wording clarity in the pr description and for general performance review and bugs made by me. it was not used for algorithm or code . since in ci pipeline documentation issues which I totally forgot was missing I am actively fixing that like using ceil(len/2) as per the definition. I am eager to explain any design or choices in detail

anurag-mds commented:
I think using frequency-based mode for floats can be misleading.

But making HSM the default for AbstractFloat might break cases where floats are actually categories or rounded values.

Maybe a keyword like discrete=false or a dispatch mode(x, HSM()) works better. I can change the API like that if the maintainers think it’s right.

anurag-mds commented:
Take a look at this:
<img width="1320" height="480" alt="image" src="https://github.com/user-attachments/assets/97da7761-faee-427e-8849-c9cdb577a7f2" />

Standard mode is unstable for continuous data  it returns different values depending on the order of elements. hsm_mode consistently identifies the true peak, as in this example: 1000 samples around 5 mixed with uniform noise. This shows why HSM should remain a separate function instead of replacing mode for floats.

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +204,99 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+
+See also: [`mode`](@ref), [`modes`](@ref)
+"""
+function hsm_mode(a::AbstractVector{T}) where T <: Real
```
You could allow any iterator given that `sort` is called immediately anyway.

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +204,99 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    len == 0 && throw(ArgumentError("hsm_mode is not defined for empty vectors"))
+    len == 1 && return a[1]
+
+    while (len > 2)
```
Avoid parentheses in this case -same with `if` below.

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +204,99 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+                best_start = i
+            end
+        end
+        a = collect(view(a, best_start:(best_start+win_size-1)))
```
Is `collect` really needed??

anurag-mds commented:
@nalimilan Does everything seems good?

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,45 @@ end
 
 Return the mode (most common number) of an array, optionally
 over a specified range `r` or weighted via a vector `wv`.
 If several modes exist, the first one (in order of appearance) is returned.
```
Does this sentence apply to `method=:halfsample`? Otherwise, move it to the description of `:default`.

nalimilan commented on src/scalarstats.jl:
````diff
@@ -46,14 +46,45 @@ end
+  For weighted mode calculations, only `:default` is supported.
+
+# Examples
+```julia
````
Use ```jldoctest and include `julia>` and the output like in the REPL (see other examples and Documenter.jl manual).

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,45 @@ end
+    if method == :halfsample
+        return _hsm_mode(a)
+    elseif method != :default
+        throw(ArgumentError("method must be :default or :halfsample, got :$method"))
```
```suggestion
        throw(ArgumentError("`method` must be `:default` or `:halfsample`, got `:$method`"))
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,45 @@ end
+function mode(a::AbstractArray{T}, r::UnitRange{T}; method::Symbol=:default) where T<:Integer
+    if method == :halfsample
+        return _hsm_mode(a)
+    elseif method != :default
```
Better have two internal methods, `_hsm_mode` and `_frequency_mode`, and call them from here.

nalimilan commented on src/scalarstats.jl:
```diff
@@ -161,7 +197,12 @@ function modes(a)
+        throw(ArgumentError("method=:halfsample is not supported for weighted mode"))
+    elseif method != :default
+        throw(ArgumentError("method must be :default or :halfsample, got :$method"))
+    end
```
Here you can just avoid allowing the `method` argument. This is reflected in the docstring, and it's not super useful to allow it if only one value is supported.

nalimilan commented on src/scalarstats.jl:
```diff
@@ -183,7 +224,12 @@ function mode(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
 end
 
-function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+function modes(a::AbstractVector, wv::AbstractWeights{T}; method::Symbol=:default) where T <: Real
```
Same here.

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +250,45 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+#   Mode: Half-Sample Mode
+#
+#############################
+
```
This kind of formatting seems to be used to differentiate different families of functions, but here HSM is just a variant of `mode`.
```suggestion
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +250,45 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+# Helper function for half-sample mode algorithm
+function _hsm_mode(a)
+    # Check for NaN
+    any(isnan, a) && throw(ArgumentError("mode with method=:halfsample does not support NaN values"))
```
```suggestion
    any(isnan, a) &&
        throw(ArgumentError("`mode` with `method=:halfsample` does not support `NaN` values"))
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +250,45 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    any(isnan, a) && throw(ArgumentError("mode with method=:halfsample does not support NaN values"))
+    
+    # Sort the data (allows any iterator)
+    a = sort(collect(a))
```
```suggestion
    a = sort!(collect(a))
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +250,45 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    a = sort(collect(a))
+    
+    len = length(a)
+    len == 0 && throw(ArgumentError("mode with method=:halfsample is not defined for empty collections"))
```
That's also the case for the other method:
```suggestion
    len == 0 && throw(ArgumentError("mode is not defined for empty collections"))
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +250,45 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    len == 2 && return middle(a[1], a[2])
+
+    while len > 2
+        win_size = cld(len, 2)  # Use ceiling division for correct half-sample
```
Not sure this comment adds anything (of course we want to do the correct thing ;-).
```suggestion
        win_size = cld(len, 2)
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,53 @@ wv = weights([0.1:0.1:0.7; 0.1])
+@test mode([10.0], method=:halfsample) ≈ 10.0
+@test mode([1.0, 2.0, 3.0], method=:halfsample) ≈ 1.5
+@test mode([1.0, 2.0, 3.0, 4.0, 5.0], method=:halfsample) ≈ 1.5
+@test isapprox(mode([1.0097, 1.0054, 1.003212, 1.0231, 1.344, 1.00003], method=:halfsample), 1.00431, atol=1e-4)
```
Can you use a higher precision? Better copy the exact value that is returned, and use `==` too.

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,53 @@ wv = weights([0.1:0.1:0.7; 0.1])
 
+## mode with method=:halfsample (half-sample mode)
+
+@test mode([1, 2, 2, 3, 4, 4, 4, 5], method=:halfsample) ≈ 4.0
```
AFAICT the value will always be the same under IEEE float arithmetic, so you can use `==` here and below?

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,53 @@ wv = weights([0.1:0.1:0.7; 0.1])
+@test mode([1.0, 1.0, 1.0], method=:halfsample) ≈ 1.0  # all identical
+@test mode([1, 1, 1, 2, 2, 100], method=:halfsample) ≈ 1.0  # outlier: finds densest region
+@test mode(Int32[1, 2, 3], method=:halfsample) isa Float64  # type promotion
+@test mode(Float32[1, 2, 3], method=:halfsample) isa Float32  # type preservation
```
This one seems redundant with the `Float32` test above. Test `Int32` instead? could also be grouped with other type tests above.

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,53 @@ wv = weights([0.1:0.1:0.7; 0.1])
+@test mode(Float32[1, 2, 3], method=:halfsample) isa Float32  # type preservation
+
+# Test overflow protection with middle()
+@test mode([typemax(Int64)-1, typemax(Int64)], method=:halfsample) ≈ float(typemax(Int64)) - 0.5
```
```suggestion
@test mode([typemax(Int64)-1, typemax(Int64)], method=:halfsample) == typemax(Int64) - 0.5
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,45 @@ end
 If several modes exist, the first one (in order of appearance) is returned.
+
+The `method` keyword argument allows selecting different mode estimation methods:
+- `:default`: Frequency-based mode (counts occurrences). This is the standard mode
```
Call this `:frequency` or `:counts`? The fact that it's the default doesn't explain what it does.

nalimilan replied:
Not addressed.

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,45 @@ end
+The `method` keyword argument allows selecting different mode estimation methods:
+- `:default`: Frequency-based mode (counts occurrences). This is the standard mode
+  that returns the most frequently occurring value in the data.
+- `:halfsample`: Half-sample mode (HSM), a robust density-based estimator for continuous data.
```
Can you cite a reference describing the method in a `# References` section?

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,53 @@ wv = weights([0.1:0.1:0.7; 0.1])
+@test mode([1, 2, 2, 3, 4, 4, 4, 5], 1:5, method=:halfsample) ≈ 4.0  # Test halfsample with range
+@test_throws ArgumentError mode([1, 2, 2, 3, 4, 4, 4, 5], 1:5, method=:invalid)
+
+
```
Could you add a test with a longer vector, just to make sure the general case is well tested?

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,53 @@ wv = weights([0.1:0.1:0.7; 0.1])
+@test mode([1.0, 1.0, 1.0, 10.0], method=:halfsample) ≈ 1.0
+@test mode([-5.0, -3.0, -1.0, 0.0, 1.0, 3.0, 5.0], method=:halfsample) ≈ -0.5
+@test_throws ArgumentError mode(Float64[], method=:halfsample)
+@test mode([1.0, 5.0], method=:halfsample) ≈ 3.0  # two elements: returns midpoint
```
Put these with the test for a single value vector.

nalimilan reviewed (commented):
Thanks!

anurag-mds commented:
Thanks @nalimilan for the thorough review. I agree with the remaining points (doc wording, method naming, references, and test adjustments). I’m addressing them now and will push a final cleanup commit so that
 everything is consolidated.

anurag-mds commented:
@nalimilan I have made necessary changes if anything to fix, modify or add do let me know!
Thanks a lot for the guidance

anurag-mds commented:
@nalimilan What do you think?

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,53 @@ end
+"""
+function mode(a::AbstractArray{T}, r::UnitRange{T}; method::Symbol=:default) where T<:Integer
+    if method == :halfsample
+        return _hsm_mode(a)
```
This isn't correct, `r` shouldn't be ignored. Do you think you could implement it? Otherwise throw an error and document it.

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,55 @@ wv = weights([0.1:0.1:0.7; 0.1])
+# Type tests
+@test mode([1, 2, 3], method=:halfsample) isa Float64
+@test mode(Int32[1, 2, 3], method=:halfsample) isa Float64  # type promotion
+@test mode([1.0f0, 2.0f0], method=:halfsample) isa Float32
```
Redundant with next one:
```suggestion
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,55 @@ wv = weights([0.1:0.1:0.7; 0.1])
+# Test mode with range and method parameter
+@test mode([1, 2, 2, 3, 4, 4, 4, 5], 1:5, method=:default) == 4
+@test mode([1, 2, 2, 3, 4, 4, 4, 5], 1:5, method=:halfsample) == 4.0  # Test halfsample with range
+@test_throws ArgumentError mode([1, 2, 2, 3, 4, 4, 4, 5], 1:5, method=:invalid)
```
Test with `2:4` to actually test that argument.

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,55 @@ wv = weights([0.1:0.1:0.7; 0.1])
+# Test weighted mode (no method parameter for weighted)
+@test mode([1, 2, 3], weights([0.1, 0.6, 0.3])) == 2
+@test modes([1, 2, 3], weights([0.1, 0.6, 0.3])) == [2]
+
```
This must already be tested elsewhere.
```suggestion
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +248,40 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    
+    len = length(a)
+    len == 0 && throw(ArgumentError("mode is not defined for empty collections"))
+    len == 1 && return float(a[1])
```
Actually we have a `middle` method precisely to ensure we return the right type even with a single argument:
```suggestion
    len == 1 && return middle(a[1])
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,55 @@ wv = weights([0.1:0.1:0.7; 0.1])
+## mode with method=:halfsample (half-sample mode)
+
+@test mode([10.0], method=:halfsample) ≈ 10.0
+@test mode([1.0, 5.0], method=:halfsample) ≈ 3.0  # two elements: returns midpoint
```
```suggestion
@test mode([10], method=:halfsample)::Float64 == 10
@test mode([1, 5], method=:halfsample)::Float64 = 3  # two elements: returns midpoint
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,55 @@ wv = weights([0.1:0.1:0.7; 0.1])
+@test mode([1.0, 5.0], method=:halfsample) ≈ 3.0  # two elements: returns midpoint
+@test_throws ArgumentError mode(Float64[], method=:halfsample)
+
+@test mode([1, 2, 2, 3, 4, 4, 4, 5], method=:halfsample) == 4.0
```
```suggestion
@test mode([1, 2, 2, 3, 4, 4, 4, 5], method=:halfsample)::Float64 == 4
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -108,7 +147,12 @@ function modes(a::AbstractArray{T}, r::UnitRange{T}) where T<:Integer
+        return _hsm_mode(a)
+    elseif method != :default
+        throw(ArgumentError("`method` must be `:default` or `:halfsample`, got `:$method`"))
+    end
```
Given that the two methods don't return the same type when inputs are integers, we need to inline this function so that the compiler knows which function is called inside the caller function. (Same above BTW.)
```suggestion
@inline function mode(a; method::Symbol=:default)
    if method === :halfsample
        return _hsm_mode(a)
    elseif method === :default
        return _frequency_mode(a)
    else
        throw(ArgumentError("`method` must be `:default` or `:halfsample`, got `:$method`"))
    end
end

# Helper function for default mode algorithm
function _frequency_mode_mode(a)
```
```

nalimilan commented on test/scalarstats.jl:
To check that there's no inference regression in the default case:
```suggestion
@test @inferred(mode(d1)) == 3
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,55 @@ wv = weights([0.1:0.1:0.7; 0.1])
 @test_throws ArgumentError modes([1, 2, 3], weights([0.1, 0.3]))
 
+## mode with method=:halfsample (half-sample mode)
+
```
```suggestion
@test mode([1], method=:default) == 1
@test mode((x for x in [1]), method=:default) == 1
my_hsm_mode(x) = mode(x, method=:default)
@test inferred(my_hsm_mode([1])) == 1
my_frequency_mode(x) = mode(x, method=:halfsample)
@test inferred(my_frequency_mode([1])) == 1

```

anurag-mds commented:
Everything Noted.
I will make sure that all issues are properly addressed...

anurag-mds commented:
Rebasing took more time Is everything good @nalimilan ?

anurag-mds commented:
Is there anything left @nalimilan ?

devmotion commented on src/scalarstats.jl:
````diff
@@ -46,14 +46,58 @@ end
+```
 """
-function mode(a::AbstractArray{T}, r::UnitRange{T}) where T<:Integer
+@inline function mode(a; method::Symbol=:frequency)
````
Why `@inline`?

```suggestion
function mode(a; method::Symbol=:frequency)
```

devmotion commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,58 @@ end
+    elseif method === :frequency
+        return _frequency_mode(a)
+    else
+        throw(ArgumentError("`method` must be `:frequency` or `:halfsample`, got `:$method`"))
```
```suggestion
        throw(ArgumentError(LazyString("`method` must be `:frequency` or `:halfsample`, got `:", method, "`")))
```

devmotion commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,58 @@ end
+    end
+end
+
+@inline function mode(a::AbstractArray{T}, r::UnitRange{T}; method::Symbol=:frequency) where T<:Integer
```
Is it necessary that the types are identical?
```suggestion
function mode(a::AbstractArray{T}, r::UnitRange{T}; method::Symbol=:frequency) where T<:Integer
```

devmotion commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,58 @@ end
+    elseif method === :frequency
+        return _frequency_mode(a, r)
+    else
+        throw(ArgumentError("`method` must be `:frequency` or `:halfsample`, got `:$method`"))
```
```suggestion
        throw(ArgumentError(LazyString("`method` must be `:frequency` or `:halfsample`, got `:", method, "`")))
```

devmotion commented on src/scalarstats.jl:
```diff
@@ -204,6 +248,43 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+
+Internal implementation of the Half-Sample Mode (HSM) estimator.
+Do not call directly; use `mode(a, method=:halfsample)` instead.
+"""
```
IMO internal functions should not have docstrings, only code comments:

```suggestion
# Internal implementation of the Half-Sample Mode (HSM) estimator
```

devmotion commented on src/scalarstats.jl:
```diff
@@ -204,6 +248,43 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    isempty(a) && throw(ArgumentError("mode is not defined for empty collections"))
+
+    # Filter non-finite values and sort
+    filtered = sort([x for x in a if isfinite(x)])
```
```suggestion
    filtered = sort!([x for x in a if isfinite(x)])
```

devmotion commented on src/scalarstats.jl:
```diff
@@ -204,6 +248,43 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    filtered = sort([x for x in a if isfinite(x)])
+    len = length(filtered)
+
+    len == 0 && throw(ArgumentError("mode is not defined for collections with no finite values"))
```
I can't see this restriction in the other mode implementation above. So maybe that is only a limitation of half-sample mode?

anurag-mds replied:
Yes, it is

As you can see for Regular Frequency Mode:

Consider [Nan, Inf, -Inf]
Here,
1) It can still count frequencies
2) Nan appears once, Inf appears once
3) Returns first one

Now, for HSM Mode:

Consider the same array,
Here,
1) Filters out all non-finite values
2) Nothing left to calculate
3) MUST throw error
4) Cannot find a cluster with no data!

nalimilan replied:
Then the error message should be more explicit:
```suggestion
    len == 0 && throw(ArgumentError("mode with `method=:halfsample` is not defined " *
                                    "for collections with no finite values"))
```

anurag-mds commented:
I've addressed the final performance suggestions (sort! and LazyString). Ready for final review.
@nalimilan @devmotion

anurag-mds commented:
Can you please review the updated changes now?

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,58 @@ end
+
+- `:halfsample`: Half-sample mode (HSM). A robust estimator of the mode for
+  continuous data. Repeatedly finds the contiguous half-sample with the smallest
+  range until ≤ 2 points remain, then returns their midpoint. The return value
```
```suggestion
  range until at most 2 points remain, then returns their midpoint. The return value
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,58 @@ end
-Return the mode (most common number) of an array, optionally
-over a specified range `r` or weighted via a vector `wv`.
-If several modes exist, the first one (in order of appearance) is returned.
+Return the mode (most common value) of `a`.
```
```suggestion
Return the mode (most common value) of `a`, optionally
over a specified range `r` or weighted via a vector `wv`.
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,58 @@ end
+    end
+end
+
+# The original frequency-based implementation, now a private helper
```
Future readers won't care about what existed before this PR.
```suggestion
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -107,8 +151,8 @@ function modes(a::AbstractArray{T}, r::UnitRange{T}) where T<:Integer
 
-# compute mode over arbitrary iterable
-function mode(a)
+# The original frequency-based implementation, now a private helper
```
```suggestion
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,58 @@ end
+  continuous data. Repeatedly finds the contiguous half-sample with the smallest
+  range until ≤ 2 points remain, then returns their midpoint. The return value
+  is always a floating-point number and may not be an element of `a`.
+  This is appropriate for continuous data and is robust to outliers.
```
Already said above. And need to say that some arguments are not allowed.
```suggestion
  This method currently does not support `r` and `wv` arguments.
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,58 @@ end
+  range until ≤ 2 points remain, then returns their midpoint. The return value
+  is always a floating-point number and may not be an element of `a`.
+  This is appropriate for continuous data and is robust to outliers.
+  See: Bickel (2002), Robertson & Cryer (1974).
```
Give full references, otherwise they are impossible to find.

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,58 @@ end
 
 # compute mode, given the range of integer values
```
```suggestion
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +248,38 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+
+    # Iteratively find the half-sample with the smallest range
+    while len > 2
+        half = ceil(Int, len / 2)
```
```suggestion
        half = cld(len, 2)
```
Right?

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +248,38 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+                best_i = i
+            end
+        end
+        filtered = filtered[best_i:(best_i + half - 1)]
```
In previous versions you used a view, why did you stop?

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,50 @@ wv = weights([0.1:0.1:0.7; 0.1])
+
+# Check no type inference regression on the default/frequency path
+my_frequency_mode(x) = mode(x, method=:frequency)
+@test @inferred(my_frequency_mode([1])) == 1
```
```suggestion
@test @inferred(mode([1]))::Int == 1

my_frequency_mode(x) = mode(x, method=:frequency)
@test @inferred(my_frequency_mode([1]))::Int == 1
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,50 @@ wv = weights([0.1:0.1:0.7; 0.1])
+## mode with method=:halfsample (half-sample mode)
+@test mode([10], method=:halfsample)::Float64 == 10.0
+@test mode([1, 5], method=:halfsample)::Float64 == 3.0         # midpoint of two elements
+@test mode([1.0, 5.0], method=:halfsample) ≈ 3.0
```
This is exactly the same as the previous case.
```suggestion
@test mode([1.0, 5.0], method=:halfsample) == 3.0
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,50 @@ wv = weights([0.1:0.1:0.7; 0.1])
+@test mode([1, 5], method=:halfsample)::Float64 == 3.0         # midpoint of two elements
+@test mode([1.0, 5.0], method=:halfsample) ≈ 3.0
+@test mode([1, 2, 2, 3, 4, 4, 4, 5], method=:halfsample)::Float64 == 4.0
+@test mode([1.0, 1.1, 1.2, 5.0, 5.1], method=:halfsample) ≈ 1.1  atol=0.1
```
`atol=0.1` is enormous. Isn't this enough?
```suggestion
@test mode([1.0, 1.1, 1.2, 5.0, 5.1], method=:halfsample) ≈ 1.1
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,50 @@ wv = weights([0.1:0.1:0.7; 0.1])
+@test mode([1.0, 1.1, 1.2, 5.0, 5.1], method=:halfsample) ≈ 1.1  atol=0.1
+
+# Robustness to outliers
+@test abs(mode([1.0, 1.05, 1.1, 1.15, 100.0], method=:halfsample) - 1.075) < 0.2
```
Better test the exact expected return value.

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,50 @@ wv = weights([0.1:0.1:0.7; 0.1])
+@test abs(mode([1.0, 1.05, 1.1, 1.15, 100.0], method=:halfsample) - 1.075) < 0.2
+
+# Non-finite values are filtered
+@test mode([1.0, NaN, 2.0, 2.0, Inf], method=:halfsample) ≈ 2.0
```
```suggestion
@test mode([1.0, NaN, 2.0, 2.0, Inf], method=:halfsample) == 2.0
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,50 @@ wv = weights([0.1:0.1:0.7; 0.1])
+## @inferred check for default mode — no inference regression
+d1 = [1, 2, 3, 3, 4]
+@test @inferred(mode(d1)) == 3
+
```
I propose testing this above instead, with similar checks.
```suggestion
```

nalimilan reviewed (commented):
Sorry for the delay. Looks almost ready to me.

@devmotion Do you have better ideas of a name for `method=:frequency`? The definition of the mode implies frequency whatever the estimation method, but I can't find a better name...

anurag-mds commented:
@nalimilan Done applied the suggested changes
Extremely sorry for such a noisy PR
But I got to know many standards and style julia follows

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,64 @@ end
+
+- `:halfsample`: Half-sample mode (HSM). A robust estimator of the mode for
+  continuous data. Repeatedly finds the contiguous half-sample with the smallest
+  range until atmost 2 points remain, then returns their midpoint. The return value
```
```suggestion
  range until at most 2 points remain, then returns their midpoint. The return value
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -46,14 +46,64 @@ end
-    mode(a, [r])
-    mode(a::AbstractArray, wv::AbstractWeights)
+    mode(a; method=:frequency)
+    mode(a::AbstractArray{T}, r::UnitRange{T}; method=:frequency) where T<:Integer
```
```suggestion
    mode(a::AbstractArray, wv::AbstractWeights; method=:frequency)
    mode(a::AbstractArray{T}, r::UnitRange{T}; method=:frequency) where T<:Integer
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +253,38 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    len = length(filtered)
+
+    len == 0 && throw(ArgumentError("mode with `method=:halfsample` is not defined " *
+                    "for collections with no finite values"))
```
```suggestion
    len == 0 && throw(ArgumentError("mode with `method=:halfsample` is not defined " *
                                    "for collections with no finite values"))
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +253,38 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+                best_i = i
+            end
+        end
+        filtered = @view filtered[best_i:(best_i + half - 1)]
```
For type stability, better use a different name like `filteredv` and do `filteredv = @view filtered[1:end]` before the loop.

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,47 @@ wv = weights([0.1:0.1:0.7; 0.1])
 
+## mode with method=:frequency (explicit, same as default)
+@test mode([1], method=:frequency) == 1
+@test mode((x for x in [1]), method=:frequency) == 1
```
Better test with values which would give something different with the other method:
```suggestion
@test mode([1, 1, 1, 2, 3, 4], method=:frequency) == 1
@test mode((x for x in [1, 1, 1, 2, 3, 4]), method=:frequency) == 1
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,47 @@ wv = weights([0.1:0.1:0.7; 0.1])
+
+
+
+@test mode([1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 5.0, 5.1, 5.2, 5.3], method=:halfsample) == 1.15
```
```suggestion
@test mode([1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 5.0, 5.1, 5.2, 5.3], method=:halfsample) == 1.15


```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +253,38 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    isempty(a) && throw(ArgumentError("mode is not defined for empty collections"))
+
+    # Filter non-finite values and sort
+    filtered = sort!([x for x in a if isfinite(x)])
```
Actually I now realize that StatsBase generally doesn't skip infinite values silently. For example `quantile` throws an error in the presence of NaN. This is less dangerous. The advantage is also that if we change our minds later we can simply turn the error into something that works, while the contrary would be breaking.

`Inf` and `-Inf` could be allowed, right? They are sorted in the right position, and if the two final values are finite, the result will be finite too. It's similar to `quantile([1, Inf], .5)` returning `Inf`.

anurag-mds commented:
Fixed the remaining points  `at most` typo, exact test value pinned (`1.0750000000000002`), error message updated to `'for collections containing only NaN values'`, `filteredv` view used throughout for type stability, frequency tests updated to use longer vector, docstring signature added.

One change I made proactively based on your comment about:

- `Inf`: changed the filter from `isfinite` to `!isnan`, so `Inf` and `-Inf` are now allowed through.

- `mode([NaN, Inf, -Inf])` returns `NaN` rather than throwing. 

**Happy to revert this if you'd prefer the stricter behaviour.**

nalimilan commented:
> * `mode([NaN, Inf, -Inf])` returns `NaN` rather than throwing.

Please throw an `ArgumentError` instead in the presence of NaN, this seems more explicit to me and it's easier to change later if we want.

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +254,40 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+    while len > 2
+        half = cld(len, 2)
+        best_i = 1
+        best_width = filtered[half] - filtered[1]
```
Need to use `filteredv` here and below, right? I'm a bit worried that this wasn't caught by tests. Isn't this covered?

anurag-mds replied:
The reason the tests didn't catch it is that the previous test cases were too simple they didn't 'squeeze' the data enough to show the mistake. I've now updated the code to use filteredv everywhere and added a more complex `test case ([1.0, 2.0, ... 13.0])` that would have failed with the old bug.

I also switched to a strict ArgumentError for any non-finite values as you suggested. Ready for another look!

anurag-mds commented:
I also wanted to say I'm genuinely sorry for the extra work this PR has caused. When this PR was created it was my very first PR, and I'm honestly a bit embarrassed that I ended up increasing your workload instead of making it easier. I’m learning a lot about the project's standards through this process, and I really appreciate your patience and guidance in helping me get this implementation right.

nalimilan commented on src/scalarstats.jl:
````diff
@@ -46,14 +46,67 @@ end
-over a specified range `r` or weighted via a vector `wv`.
-If several modes exist, the first one (in order of appearance) is returned.
+# Examples
+```julia
````
```suggestion
```jldoctest
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +256,41 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
     return [x for (x, w) in weights if w == mw]
 end
 
+#Internal implementation of the Half-Sample Mode (HSM) estimator.
```
```suggestion
# Internal implementation of the Half-Sample Mode (HSM) estimator.
```

nalimilan commented on src/scalarstats.jl:
```diff
@@ -204,6 +256,41 @@ function modes(a::AbstractVector, wv::AbstractWeights{T}) where T <: Real
+#Internal implementation of the Half-Sample Mode (HSM) estimator.
+function _hsm_mode(a)
+    isempty(a) && throw(ArgumentError("mode is not defined for empty collections"))
+
```
```suggestion
    if !all(x -> x isa Real, a)
        throw(ArgumentError("mode with `method=:halfsample` is only defined " *
                            "for collections containing real numbers"))
    end
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,53 @@ wv = weights([0.1:0.1:0.7; 0.1])
+
+# Robustness to outliers
+@test mode([1.0, 1.01, 1.011, 100.0, 200.0], method=:halfsample) == 1.0105
+
```
```suggestion

# Non-real values throw ArgumentError
@test_throws ArgumentError mode(["a", "b", "c"], method=:halfsample)
@test_throws ArgumentError mode([1+1im, 1+2im], method=:halfsample)

```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,53 @@ wv = weights([0.1:0.1:0.7; 0.1])
+
+# Edge cases
+@test_throws ArgumentError mode(Float64[], method=:halfsample)
+
```
```suggestion
```

nalimilan commented on test/scalarstats.jl:
```diff
@@ -63,6 +63,53 @@ wv = weights([0.1:0.1:0.7; 0.1])
+@test_throws ArgumentError mode([1, 2, 2, 3, 4, 4, 4, 5], 2:4, method=:halfsample)
+@test_throws ArgumentError mode([1, 2, 2, 3, 4, 4, 4, 5], 2:4, method=:invalid)
+
+
```
```suggestion
```

nalimilan reviewed (commented):
Looks good modulo some suggestions I'm making. I'll apply them and see whether tests pass.

nalimilan committed:
Apply suggestions from code review

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

nalimilan committed:
Merge branch 'master' into hsm-mode
